File: Lintcode/167_add-two-numbers/add-two-numbers.cpp
/*
@Copyright:LintCode
@Author:   linqiao
@Problem:  http://www.lintcode.com/problem/add-two-numbers
@Language: C++
@Datetime: 17-04-04 08:53
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param l1: the first list
     * @param l2: the second list
     * @return: the sum list of l1 and l2 
     */
    ListNode *addLists(ListNode *l1, ListNode *l2) {
        // write your code here
        if (l1 == NULL) return l2;
        if (l2 == NULL) return l1;
        ListNode *head = NULL;
        ListNode *temp = head;
        int carry = 0;
        while (l1 != NULL && l2 != NULL) {
            if (l1->val + l2->val + carry >= 10) {
                if (head == NULL) {
                    head = new ListNode(l1->val+l2->val+carry-10);
                    temp = head;
                    carry = 1;
                }
                else {
                    ListNode *temp2 = new ListNode(l1->val+l2->val+carry-10);
                    temp->next = temp2;
                    temp = temp->next;
                    carry = 1;
                }
            }
            else {
                if (head == NULL) {
                    head = new ListNode(l1->val+l2->val+carry);
                    temp = head;
                    carry = 0;
                }
                else {
                    ListNode *temp2 = new ListNode(l1->val+l2->val+carry);
                    temp->next = temp2;
                    temp = temp->next;
                    carry = 0;
                }
            }
            l1 = l1->next;
            l2 = l2->next;
        }
        if (l1 == NULL) {
            while (l2 != NULL) {
                if (l2->val+carry >= 10) {
                    temp->next = new ListNode(l2->val+carry-10);
                    temp = temp->next;
                    carry = 1;
                }
                else {
                    temp->next = new ListNode(l2->val+carry);
                    temp = temp->next;
                    carry = 0;
                }
                l2 = l2->next;                
            }
        }
        else if (l2 == NULL) {
            while (l1 != NULL) {
                if (l1->val+carry >= 10) {
                    temp->next = new ListNode(l1->val+carry-10);
                    temp = temp->next;
                    carry = 1;
                }
                else {
                    temp->next = new ListNode(l1->val+carry);
                    temp = temp->next;
                    carry = 0;
                }
                l1 = l1->next;                
            }
        }
        if (carry == 1) {
            temp->next = new ListNode(1);
        }
        return head;
    }
};

File: LeetCode/Add Two Numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		int carry = 0;
		ListNode l = new ListNode(0);
		ListNode t = l;
		while (l1 != null && l2 != null) {
			int s = l1.val+l2.val+carry;
			if (s < 10) {
				t.next = new ListNode(s);
				carry = 0;
				t = t.next;
			}
			else {
				t.next = new ListNode(s-10);
				carry = 1;
				t = t.next;
			}
            l1 = l1.next;
            l2 = l2.next;
		}
		ListNode m = null;
		if (l1 == null) m = l2;
		else if (l2 == null) m = l1;
		while (m != null) {
			int s = m.val+carry;
			if (s < 10) {
				t.next = new ListNode(s);
				carry = 0;
				t = t.next;
			}
			else {
				t.next = new ListNode(s-10);
				carry = 1;
				t = t.next;
			}
            m = m.next;
		}
		if (carry == 1) t.next = new ListNode(1);
        l = l.next;
		return l;
    }
}